modin/test/storage_formats/pandas/test_internals.py

```diff
@@ -110,3 +110,79 @@ def func_to_apply(partition, row_internal_indices, col_internal_indices, item):
     md_df._query_compiler._modin_frame = new_modin_frame
 
     df_equals(md_df, pd_df)
+
+
+small_dfs = [
+    pd.DataFrame(
+        [[i + j for j in range(0, 1000)]],
+        columns=[f"col{j}" for j in range(1, 1001)],
+        index=pd.Index([i - 1]),
+    )
+    for i in range(1, 100001, 1000)
+]
+large_df = pd.DataFrame(
+    [[i + j for j in range(1, 1000)] for i in range(0, 100000, 1000)],
+    columns=[f"col{j}" for j in range(1, 1000)],
+    index=pd.Index(list(range(0, 100000, 1000))),
+)
+
+
+@pytest.mark.skipif(
+    Engine.get() not in ("Dask", "Ray"),
+    reason="Rebalancing partitions is only supported for Dask and Ray engines",
+)
+@pytest.mark.parametrize(
+    "large_df,col_length",
+    [
+        (pd.concat(small_dfs), 100),
+        (pd.concat([pd.concat(small_dfs)] + small_dfs[:3]), 103),
+        (pd.concat([large_df] + small_dfs[:3]), 103),
+    ],
+)
+def test_rebalance_partitions(large_df, col_length):
+    large_modin_frame = large_df._query_compiler._modin_frame
+    assert large_modin_frame._partitions.shape == (
+        NPartitions.get(),
+        NPartitions.get(),
+    ), "Partitions were not rebalanced after concat."
+    assert all(
+        isinstance(ptn, large_modin_frame._partition_mgr_cls._column_partitions_class)
+        for ptn in large_modin_frame._partitions.flatten()
+    )
+    # The following check tests that we can correctly form full-axis virtual partitions
+    # over the orthogonal axis from non-full-axis virtual partitions.
+
+    def col_apply_func(col):
+        assert len(col) == col_length, "Partial axis partition detected."
```

Comment:
Why do we need this `assert`? Why isn't it enough to check that the column-wise apply does what pandas would do?

Reply:
We want to ensure that this is a full-axis partition, and not just a partial axis partition - the function as I've written it will run correctly even if it is a partial axis partition, so I just thought this assert would be a good way to keep the function simple, but still test that it is full axis.

Reply:
I think this test should check for functional correctness. We already check that we get the `apply` result correct when the starting partitions are non-full-axis virtual partitions. I don't think it's right to check for internal implementation details here.

Reply:
Hmm - I'm not sure that I agree with that - this file is called `test_internals` which would imply its testing internal details, and it makes sense to ensure that the internal partitioning code didn't incorrectly make a full axis partition.

Reply:
I don't think it's appropriate to test any internal detail here just because it's `test_internals`. Why not do something like have this function return `col[::-1]`? That way we can be pretty confident that modin is really applying the function across the full axis. I think testing the intermediate representations during the function call is outside the scope of this particular test case.

Reply:
I could do that, but then I would have to go into each partition to determine what the last row is and then double check the output, so this feels "less intrusive" open to other suggestions though! Perhaps the solution may be to add unit testing for partition layer stuff into the codebase like @pyrito suggested, and move this code to there, so that `test_internals` just tests the overall internals as its supposed to, and implementation details can be tested at that level?